tfhe/src/integer/gpu/server_key/radix/scalar_mul.rs

```diff
@@ -34,64 +38,133 @@ impl CudaServerKey {
-        scalar: u64,
-        stream: &CudaStream,
-    ) -> CudaUnsignedRadixCiphertext {
+    pub fn unchecked_scalar_mul<Scalar, T>(&self, ct: &T, scalar: Scalar, stream: &CudaStream) -> T
+    where
+        Scalar: ScalarMultiplier + DecomposableInto<u8> + CastInto<u64>,
+        T: CudaIntegerRadixCiphertext,
+    {
         let mut result = unsafe { ct.duplicate_async(stream) };
-        self.unchecked_small_scalar_mul_assign(&mut result, scalar, stream);
+        self.unchecked_scalar_mul_assign(&mut result, scalar, stream);
         result
     }
 
     /// # Safety
     ///
     /// - `stream` __must__ be synchronized to guarantee computation has finished, and inputs must
     ///   not be dropped until stream is synchronised
-    pub unsafe fn unchecked_small_scalar_mul_assign_async(
+    pub unsafe fn unchecked_scalar_mul_assign_async<Scalar, T>(
         &self,
-        ct: &mut CudaUnsignedRadixCiphertext,
-        scalar: u64,
+        ct: &mut T,
+        scalar: Scalar,
         stream: &CudaStream,
-    ) {
-        match scalar {
-            0 => {
-                ct.as_mut().d_blocks.0.d_vec.memset_async(0, stream);
-            }
-            1 => {
-                // Multiplication by one is the identity
+    ) where
+        Scalar: ScalarMultiplier + DecomposableInto<u8> + CastInto<u64>,
+        T: CudaIntegerRadixCiphertext,
+    {
+        if scalar == Scalar::ZERO {
+            ct.as_mut().d_blocks.0.d_vec.memset_async(0, stream);
+            return;
+        }
+
+        if scalar == Scalar::ONE {
+            return;
+        }
+
+        if scalar.is_power_of_two() {
+            // Shifting cost one bivariate PBS so its always faster
+            // than multiplying
+            self.unchecked_scalar_left_shift_assign_async(ct, scalar.ilog2() as u64, stream);
+            return;
+        }
+        let ciphertext = ct.as_mut();
+        let num_blocks = ciphertext.d_blocks.lwe_ciphertext_count().0;
+        let msg_bits = self.message_modulus.0.ilog2() as usize;
+        let decomposer = BlockDecomposer::with_early_stop_at_zero(scalar, 1).iter_as::<u8>();
+
+        // We don't want to compute shifts if we are not going to use the
+        // resulting value
+        let mut has_at_least_one_set = vec![0u64; msg_bits];
+        for (i, bit) in decomposer.collect_vec().iter().copied().enumerate() {
+            if bit == 1 {
+                has_at_least_one_set[i % msg_bits] = 1;
             }
-            _ => {
-                let lwe_dimension = ct.as_ref().d_blocks.lwe_dimension();
-                let lwe_ciphertext_count = ct.as_ref().d_blocks.lwe_ciphertext_count();
+        }
 
-                stream.small_scalar_mult_integer_radix_assign_async(
+        let decomposer = BlockDecomposer::with_early_stop_at_zero(scalar, 1).iter_as::<u8>();
+        let scalar64 = decomposer
+            .collect_vec()
+            .iter()
+            .map(|&x| x as u64)
+            .collect_vec();
```

Comment:
`decomposer.iter_as::<u64>().collect::<Vec<_>>();`

Reply:
I get this error:
```100 |         let scalar64 = decomposer.iter_as::<u64>().collect::<Vec<_>>();
    |                                   ^^^^^^^ method not found in `impl Iterator<Item = u8>`
```

Reply:
ok, because the line above should be `let decomposer = BlockDecomposer::with_early_stop_at_zero(scalar, 1);`

so we could merge the two 
```rust
let decomposed_scalar = BlockDecomposer::with_early_stop_at_zero(scalar, 1)
    .iter_as::<u64>()
   .collect::<Vec<_>>();
```